### DISCUSSION TITLE
tutorials/io/data_paths

### DISCUSSION THREAD

**Discussion by giscus** [godotengine/godot-docs-user-notes#106] [User-contributed notes]
# tutorials/io/data_paths

This page explains how file paths work inside Godot projects. You will learn how to access paths in your projects using the res:// and user:// notations, and where Godot stores project and editor f...

https://docs.godotengine.org/en/stable/tutorials/io/data_paths.html

**Comment by etherealxx** [comment]
Use `String.path_join()` to join two file paths, more about that [here](https://docs.godotengine.org/en/stable/classes/class_string.html#class-string-method-path-join). Basically the gdscript equivalent of `os.path.join()` in python (with slightly different usage).
I wrote this here since this is the first page that shows up when i search for "Godot file path join" on web search, and most likely relevant.

**Reply by Calinou** [reply]
I've opened a PR to document this:

- https://github.com/godotengine/godot-docs/pull/9908

**Comment by bertodelrio256** [comment]
It is also unclear how to use paths in an exported project. use of DirAccess and FileAccess do not work on res:// paths in an exported project. So how would ResourceLoader.Load() get the paths of files in res:// in an exported project, if the paths are not already hard coded? ie... i need to loop through all files in a folder in res:// and load them.

**Reply by bertodelrio256** [reply]
ok so it looks like DirAccecss does work on exported projects. not sure why? it does return the paths as "res://file.example" which would be invalid so not sure how ResourceLoader figures out the correct path from that? anyway this code is valid on  the exported project:

```cs
public override void _Ready()
	{
		foreach (string qr in DirAccess.Open("res://db/items/").GetFiles())
		{
			item_db.Add(ResourceLoader.Load<Item>("res://db/items/" + qr));
		}

		foreach (string qr in DirAccess.Open("res://db/skills/").GetFiles())
		{
			skill_db.Add(ResourceLoader.Load<Skill>("res://db/skills/" + qr));
		}

		foreach (string qr in DirAccess.Open("res://db/quest/").GetFiles())
		{
			questdb.Add(ResourceLoader.Load<Quest>("res://db/quest/" + qr));
		}
	}
```

**Reply by rozab** [reply]
I was having trouble with this too. It looks like res:// can and should be queried in exported projects, but some files are omitted depending on the filetype:

- some built in files like .gd and .tres seem untouched

- According to [this post](https://forum.godotengine.org/t/exported-game-does-not-include-neccesary-resources/51541/2), .tcsn files likely won't be listed. Instead there will be a .tcsn.remap file, and you can access the scene by stripping the suffix from this. This doesn't appear to be documented anywhere.

- [As per the DirAccess docs](https://docs.godotengine.org/en/stable/classes/class_diraccess.html#class-diraccess-method-get-files), any imported resources like .pngs will not be listed, since the original files aren't actually exported. Instead, search for the corresponding .import file and strip the suffix.

This is pretty confusing since I guess the res:// path you pass to `load` doesn't actually represent a path that exists in res://, it's more like a key `ResourceLoader` uses to identify the resource.

It's all so awkard, I feel it would be much better if `ResourceLoader` just had a `get_files_at` equivalent which uses the same logic as `ResourceLoader.exists`.

**Reply by Calinou** [reply]
This should be addressed by https://github.com/godotengine/godot/pull/96590.

**Comment by hogdanish** [comment]
The docs here say that Godot complies with the XDG Base Directory Specification on all platforms because you can override environment variables to change the editor and project data paths, but there's no indication of what those environment variables actually are. 

They certainly aren't XDG_CONFIG_HOME like you might expect. I have all the base dirs set globally in my shell on macOS, yet launching godot from the same shell still results in `~/Library/Application Support/Godot/` being used.

**Reply by Calinou** [reply]
The docs actually seem wrong here, as `XDG_DATA_HOME`, `XDG_CONFIG_HOME` and `XDG_CACHE_HOME` only appear in Linux-related files in Godot's source code (outside of comments). Therefore, there's no way it can work on Windows and macOS…

Pull request opened:

- https://github.com/godotengine/godot-docs/pull/10009

**Reply by hogdanish** [reply]
Thanks, that makes sense then. 

I will add: using XDG base dirs on macOS is a reasonably common practice, as many cross-platform *nix applications prefer XDG directories if the variables are set in the user's environment or if a user places config files the dirs. Plenty more use them as secondary or alternative config directories, and many shell & core utils even exclusively use XDG dirs. After all, macOS is Unix, so they're pretty compatible as far as far as overall directory structure and config files go. 

A common use case for this is tracking and syncing your dotfiles for multi-platform applications across macOS and Linux machines. It's a seamless process as long as the software saves its configurations to $XDG_CONFIG_HOME on both platforms. Sadly, aside from a few app store apps, Godot is currently the only application I regularly work with on macOS that requires use of symlinks to achieve this.

I might make a feature request for this, or even look into a PR if it's a non-trivial change- even as a hidden option, I think allowing optional use of XDG dirs would be a nice quality of life feature for cross-platform users.

Edit: https://github.com/godotengine/godot-proposals/issues/10838

**Comment by tcibils** [comment]
When in doubt of where your files are, use print(OS.get_data_dir()) !

**Comment by SememeS** [comment]
## re-quote:
> No need for any of the path retrieval. In release, use "./" which starts relative from the binary.

[Exporting pcks comment link](https://github.com/godotengine/godot-docs-user-notes/discussions/13#discussioncomment-11606071)

[original reddit post link](https://www.reddit.com/r/godot/comments/zlk68u/comment/j07a384/?utm_source=share&utm_medium=web3x&utm_name=web3xcss&utm_term=1&utm_content=share_button)

**Reply by ninstar** [reply]
As of 4.4 it still seems to be the case, shouldn't this be added to the doc at this point?

**Reply by Calinou** [reply]
I don't know what "path retrieval" is referring to in this context. Also, this page is about file paths in Godot projects, not exporting PCKs for use with mods or DLCs. This is covered on a separate page: [Exporting packs, patches and mods](https://docs.godotengine.org/en/latest/tutorials/export/exporting_pcks.html)

In fact, a similar comment was cross-posted in the discussions for that page: https://github.com/godotengine/godot-docs-user-notes/discussions/13#discussioncomment-11606071

**Reply by XandaPanda42** [reply]
Using "./" as an OS path should refer to the folder that the exported executable is in. That's what they mean by path retrieval (I believe). If that works in Godot, it could be a good way to make a self contained exported project.

**Comment by Peristroff** [comment]
This stuff is super useful, I was searching inside the engine documentation about the paths and I couldn't find it, is this intended or am I just bad at searching?

**Reply by Calinou** [reply]
I don't know which class this would be the most relevant to link in. The class reference isn't intended to be a 1:1 copy of the manual; it documents the classes available in the engine.

**Comment by ShalokShalom** [comment]
I do miss a mention of the crash logs

**Reply by Calinou** [reply]
Pull request opened:

- https://github.com/godotengine/godot-docs/pull/10812

**Comment by Leroymilo** [comment]
I'm making a level editor along with my game, and my initial idea was to put user made files (they would include resource files, text files and images) next to the executable so that everything is close together, especially since users would be able to provide their own images that they'll make with other programs.
After reading this, I am affraid that this might cause issues on some systems, because I am unsure about the meaning of:
> This is because when the game is running, the project's file system will likely be read-only.

Does this refer to files inside `res://` or to anything near the executable? Is it necessary to save every generated file to `user://`? I fear that it would make my editor less user-friendly.

**Reply by jinnijinnjinn** [reply]
Working with a similar project, where I want to allow players to install texture packs to the executable folder. I think "res://" only works from within the editor, so I plan to swap it for OS.get_executable_path() before exporting.
